components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3188,39 +3215,68 @@ subroutine evaporate_sublimate_precip(qr_incld,qc_incld,nr_incld,qitot_incld,
 
    real(rtype) :: qclr, cld
 
-   ! It is assumed that macrophysics handles condensation/evaporation of qc and
-   ! that there is no condensation of rain. Thus qccon, qrcon and qcevp have
-   ! been removed from the original P3-WRF.
-
-   ! Determine temporary cloud fraction, set to zero if cloud water + ice is
-   ! very small.  This will ensure that evap/subl of precip occurs over entire
-   ! grid cell, since min cloud fraction is specified otherwise.
-   if (qc_incld + qitot_incld < 1.e-6_rtype) then
-      cld = 0._rtype
+   ! First, check that qv<=qvs. If not, unexpected supersaturation has occurred
+   ! and the assumptions of this parameterization will have been violated.
+   ! This check should eventually be commented out and/or moved to a debug-mode
+   ! feature.
+   !if ( qv-qvs .gt. 1.e-12_rtype ) then
+   !    print*
+   !    print*,'In evap_precip, qv supersaturated. qv=',qv,', qvs=',qvs,', diff=',qv-qvs
+   !    print*
+   !    call endscreamrun("qv supersaturated in evap_precip")
+   ! endif
+   
+   ! Determine temporary cloud fraction which ramps to 0 as cloud water
+   ! approaches zero. This avoids problems due to lcldm having a positive
+   ! minimum value which would result in evaporation never occurring in a
+   ! small portion of the grid. Ramping rather than thresholding is used
+   ! to preserve convergence.
+   if (qc_incld < 1.e-6_rtype) then
```

Comment:
@PeterCaldwell 

Would you explain how convergence affects here?
You chose relatively high cloud droplet mass to represent a cloudy area.

Reply:
Suppose qc_incld is infinitesimally bigger than 1e-6 at coarser resolution and drops to infinitesimally smaller than 1e-6 at higher resolution. All of a sudden the code would be handled quite differently if no ramp were present. This often causes jumps on convergence plots. I'd like to always use ramps at all thresholds but some cases are hard to handle. Does that make sense? Should I change the comment to be more clear? And yes, 1e-6 is a fairly large threshold. The fact that we're ramping instead of thresholding below that probably makes it less important, but a smaller number is worth testing.

Reply:
Thanks for the explanation.